deterministic_kernel.hpp

```diff
@@ -0,0 +1,246 @@
+    int prev_row = -1;
+    int prev_col = -1;
+    // number of rows/cols in the probability window
+    int number_of_rows = 0;
+    int number_of_columns = 0;
+    // maximum distance from center cell to outer cells
+    double max_distance;
+    std::vector<std::vector<double>> probability;
+    std::vector<std::vector<double>> probability_copy;
+    CauchyDistribution cauchy;
+    ExponentialDistribution exponential;
+    double number_of_dispersers;
+public:
+    DeterministicDispersalKernel(DispersalKernelType dispersal_kernel, 
+    							Raster<int> dispersers, double dispersal_percentage,
+    							double ew_res, double ns_res, double distance_scale = 1.0,
+    							double locator = 0.0)
+        :
+        dispersers_(dispersers),
+        cauchy(distance_scale, locator),
+        exponential(distance_scale),
+        kernel_type_(dispersal_kernel),
+        east_west_resolution(ew_res),
+        north_south_resolution(ns_res)
+    {
+    	if (kernel_type_ == DispersalKernelType::Cauchy) {
+    		max_distance = cauchy.cdf(dispersal_percentage);
+    	} else if (kernel_type_ == DispersalKernelType::Exponential) {
+    		max_distance = exponential.cdf(dispersal_percentage);
+    	} else { 
+    		//for expanding compatible kernel types
+    	}
+    	number_of_columns = ceil(max_distance / east_west_resolution) * 2 + 1;
+    	number_of_rows = ceil(max_distance / north_south_resolution) * 2 + 1;
+    	probability.resize(number_of_rows, std::vector<double>(number_of_columns));
+    	probability_copy.resize(number_of_rows, std::vector<double>(number_of_columns));
+    	mid_row = number_of_rows / 2;
+    	mid_col = number_of_columns / 2;
+    	double sum = 0.0;
+    	for ( int i = 0; i < number_of_rows; i++ ) {
+    		for ( int j = 0; j < number_of_columns; j++ ) {
+    			double distance_to_center = 
+    			sqrt(pow((abs(mid_row - i) * east_west_resolution), 2) 
+    			+ pow((abs(mid_col - j) * north_south_resolution), 2));
+    			// determine probability based on distance
+				if (kernel_type_ == DispersalKernelType::Cauchy) {
+ 					probability[i][j] = abs(cauchy.pdf(distance_to_center));
+ 				} else if ( kernel_type_ == DispersalKernelType::Exponential ) {
+ 					probability[i][j] = abs(exponential.pdf(distance_to_center));
+				}
+ 				sum += probability[i][j];
+    		}
+    	}
+    	//normalize based on the sum of all probabilities in the raster
+    	for ( int i = 0; i < number_of_rows; i++ ) {
+    		for ( int j = 0; j < number_of_columns; j++ ) {
+    			probability[i][j] /= sum;
+    		}
+    	}
+    }
+
+    /*! Generates a new position for the spread.
+     *
+     *  Creates a copy of the probability matrix to mark where dispersers are assigned.
+     *  New window created any time a new cell is selected from simulation.disperse
+     *
+     *  Selects next row/col value based on the cell with the highest probability
+     *  in the window.
+     *
+     */
+    std::tuple<int, int> spread(int row, int col)
```

Comment:
This could be a function call operator as it is in the other kernels (just copy-paste the syntax and then delete `.spread` from the place where you call it).

With how you designed your code, you are not using this kernel directly as the other kernels are used, so it is not necessary, but using the operator would be more consistent.